### PR TITLE
Pi/utf8

### LINKED ISSUE #2836: [Product Import] Importing products with special characters fails
## Description

The product import expects CSV files in UTF-8 character encoding. If the uploaded file is in a different encoding and contains characters like German umlauts, the application crashes (500 server error).

## Expected Behavior

Uploading an invalid file results in a validation error being displayed.

## Actual Behavior

The generic snail page is displayed. The log contains:
```
ArgumentError (invalid byte sequence in UTF-8):
  app/models/product_import/product_importer.rb:223:in `rows'
  app/models/product_import/product_importer.rb:245:in `build_entries'
  app/models/product_import/product_importer.rb:180:in `init_product_importer'
  app/models/product_import/product_importer.rb:38:in `initialize'
  app/controllers/admin/product_import_controller.rb:15:in `new'
  app/controllers/admin/product_import_controller.rb:15:in `import'
```

## Steps to Reproduce

1. Go to the product import, e.g. https://openfoodnetwork.de/admin/product_import
2. Select a file containing special characters in a local encoding. 
[KL test 181005 3.txt](https://github.com/openfoodfoundation/openfoodnetwork/files/2454719/KL.test.181005.3.txt) (rename to `.csv`)
3. Click "Upload"

## Context

Kirsten was testing the product import for the first German enterprises with a test file. They are pulling the data from another database.

## Severity

s3 as product import is not officially released and it affects only some users. The workaround is to use UTF8-encoded CSV files.

## Your Environment

* Version used: 58a99a7f8993cf609d930a68d82b382fccce8ea4

## Possible Fix

Rescue the `ArgumentError` and display a helpful message: "Please check the language setting of your source file and ensure it is saved in UTF-8.csv encoding".

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #2836 

Addresses issues with different encodings in spreadsheets causing errors when reading the file in product import.

#### What should we test?

When a file with unusual encoding is used with product import, an error message is displayed when it is uploaded.

#### Release notes

Added file encoding error messages to product import.

Changelog Category: Added

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
I'm not sure how to spec this one, or if it's possible to detect and convert the encoding of the file during upload...?

File: app/models/product_import/product_importer.rb [hunk 1]
```diff
@@ -226,6 +226,14 @@ def rows
       (2..@sheet.last_row).map do |i|
         @sheet.row(i)
       end
+    rescue ArgumentError => e
+      if e.message.include? 'invalid byte sequence'
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/models/product_import/product_importer.rb]: Looks like no, and neither does Ruby.

**Comment by luisramos0** [comment]
staged https://staging.openfoodfrance.org

**Comment by RachL** [comment]
hmmm I've used the file mentionned in the issue, but I think that Libre Office is somewhat re-saving the file in a correct format because I can now upload it :/ 

What kind of special character can I introduce? @Matt-Yorkley did you work with an example?

**Comment by Matt-Yorkley** [comment]
I used the file in the original issue, but I didn't re-save it.

**Comment by RachL** [comment]
@Matt-Yorkley can you tell me what can of error message I should expect? Kirsten gave me her original file so now I get an error message saying "some products do not have an enterprise defined". This message is shown usually when either you have the wrong column name or a producer in the list does not have a profile on the instance, but here I'm not in those cases.

**Comment by Matt-Yorkley** [comment]
I was working with this file from the original issue: https://github.com/openfoodfoundation/openfoodnetwork/files/2454719/KL.test.181005.3.txt

**Comment by RachL** [comment]
@Matt-Yorkley I still manage to import them correctly...
image

The only thing I can see is either A. I'm doing something wrong B. All the weird characters where in the category column and maybe french staging is not the best staging to test that? Anyway I feel I'm wasting people's time here not being able to reproduce the errror :/
can someone else have a look maybe @myriamboure @sstead ?
Here is Kirsten's file as well : 
[JA!_Artikel_Exportdatei - OFN upload test.xlsx](https://github.com/openfoodfoundation/openfoodnetwork/files/2603967/JA._Artikel_Exportdatei.-.OFN.upload.test.xlsx)

**Comment by myriamboure** [comment]
Ok I spent quite a lot of time but found how to do.
https://docs.google.com/document/d/1ioWyrCSc_KqUih5YlzOyi2N3mKubTuAYUxWRtnqy7IE/edit#

@RachL for info in libroffice to encode in what you want you have to click on "edit filter parameter" and then it gives you the option on what you want to encode. https://help.libreoffice.org/Calc/Importing_and_Exporting_CSV_Files/fr

So it seems it's not working as expected. I encoded in UTF-7 for instance, it didn't tell me then it was a wrong encoding, it displayed the message "Variant_unit_name can't be empty if unit_type is empty"... but in the file variant_unit_name is not empty, so it's unrelevant...

This is the file I used https://drive.google.com/open?id=1semMatW0iWJXpEhM7yt4L49MsFDRbJwp which is the same as @Matt-Yorkley the one you used, original Kirsten issue file. It was in text so I had to open and save as CSV and used UTF-7 encoding.

If we are doing something wrong, let us know, else there is an issue with the PR, it's not the good error message displayed, I would expect at the very beginning when I click on first button an error message saying encoding is wrong, even before I see the producer found.

**Comment by RachL** [comment]
thanks @myriamboure I also had that message in previous tests, I thought I was doing something wrong !

**Comment by Matt-Yorkley** [comment]
It worked in my dev environment :sob: 

From the reading I've done on the topic it looks like detecting file encodings is not an exact science. When you tried encoding in UTF-7, did the import work successfully?

**Comment by Matt-Yorkley** [comment]
The error will appear if the importer fails to read the file. So you're saying it's reading the file without problems in these tested scenarios?

**Comment by myriamboure** [comment]
Hum... I'm sorry @Matt-Yorkley but I'm not sure I understand... if you look at testing notes I put some screenshots so you see how the file is read I guess https://docs.google.com/document/d/1ioWyrCSc_KqUih5YlzOyi2N3mKubTuAYUxWRtnqy7IE/edit#heading=h.gxnh3o3sfsft

I don't think this issue is blocking the release of the feature though as there is a workaround and we can specify in user guide what encoding needs to be used. So @Matt-Yorkley have a look, it you don't find what the problem, I suggest we just make sure the prescribed encoding is working fine, and we can merge, and we will see the problem that emerge when it is live and used. Also, @kirstenalarsen could test on German instance once this is merged to see if the problem is fixed for Germany in the encoding case she was testing.

Does it sound like a good plan ?

**Comment by Matt-Yorkley** [comment]
Okay, so are you saying we should merge?

**Comment by myriamboure** [comment]
I have no idea @Matt-Yorkley, I suggested it as I don't see how to move forward, neither @RachL nor I manage to see the encoding error detected and helpful message displayed, even if we did encode in UTF-7 (check the testing not to see if you think it was properly encoded). So we are basically unable to know if we do something wrong, or if the code is broken and display some unadapted and confusing error message in case of a wrong encoding.
So as we are still beta testing the feature I was suggesting maybe we merge and see with countries where there is a data encoding issue if that is fixed... Maybe @mkllnk or @sauloperez who reviewed will have a clue on the best strategy to move forward ? And tell if we do something wrong as testers or if our wrong encoding fail, or if you believe the problem comes from the PR ?

**Comment by Matt-Yorkley** [comment]
It looks like it's able to read UTF-7 without any difficulty. I think to get it to fail properly we need a more obscure encoding. The file linked in #2836 was giving me the expected error when I was testing it previously.

**Comment by myriamboure** [comment]
In the testing notes @Matt-Yorkley it doesn't seem to read UTF7 properly, as "variant_unit_name" is not empty, but it's telling me it is with an error message... so is it linked to UTF7 encoding ? 
We did test with the same file as you, just renaming the file to csv... it seems it was not working, but I don't see the trace in the testing notes, is it possible to stage again so we can retest and decide what we do ? Whatever, I think the case in the release note shows that there is a problem... the error displayed is wrong and it's probably linked to the encoding, as when importing with UTF8 encoding exactly the same file we had no error message... can someone restage ?

**Comment by daniellemoorhead** [comment]
OK peeps, so there's been no action on this for over 3 weeks. I'm moving it back to the Dev Ready column, @Matt-Yorkley feel free to move it back into the doing column when you pick it up again.

Ping @myriamboure @RachL FYI

#toottoot

**Comment by myriamboure** [comment]
@daniellemoorhead I'm actually not sure what to do with that and I think we need another dev input. @mkllnk can you have a look and tell what you think we should do ? Looking at the testing notes and previous comment, do you think there is indeed a problem in the code that needs fixing ? Or shall we restage and retest but then I don't know how to do it as I did import the original file but get a wrong error message... @Matt-Yorkley says he tested with the original file and it works, but I can't reproduce. Or shall we just merge and iterate if it breaks something ? I'm lost to be honest here, I need help on how to sort this mess.

**Comment by mkllnk** [comment]
I verified on my local computer with Kirsten's test file [KL test 181005 3.txt](https://github.com/openfoodfoundation/openfoodnetwork/files/2454719/KL.test.181005.3.txt) (rename to `.csv`) that the master branch fails and Matt's branch gives the correct error message. I staged it on the Australian staging server and got the right error message.

screenshot from 2019-01-09 14-57-40

I imported one product using a valid CSV file and that worked. :heavy_check_mark:

So I think this pull request is fixing the related issue.

@myriamboure I also tried the other spreadsheet and got a similar issue to you. I can't read French, so I'm not sure if I got the same error. I think that's a new issue. I get the same on the master branch.

screenshot from 2019-01-09 15-33-25

Trying this CSV file:
```csv
supplier,sku,name,display_name,category,units,unit_type,variant_unit_name,price,on_hand,available_on,on_demand,shipping_category,tax_category
JA,453586005," Schokolade, massiv",Kaffee-Sahne,Süßwaren,1,,Item,1.59,0,,1,,MwST-7
JA,453586002," Schokolade, massiv",Rahm-Mandel,Süßwaren,1,,Item,1.59,0,,1,,MwST-7
JA,453586003," Schokolade, massiv",Trauben-Nuss,Süßwaren,1,,Item,1.59,0,,1,,MwST-7
JA,231843001, Schmelzkäse-Scheiben,Allgäuer,Frischwaren,1,,Item,1.19,0,,1,,MwST-7
JA,231843002, Schmelzkäse-Scheiben,Chester,Frischwaren,1,,Item,1.19,0,,1,,MwST-7
JA,231843003, Schmelzkäse-Scheiben,Toast,Frischwaren,1,,Item,1.19,0,,1,,MwST-7
JA,546274, Müllbeutel,mit Tragegriff,Haushaltwaren,25,L,,0.69,0,,1,,MwST-7
JA,369466, Müllbeutel ,mit Zugband,Haushaltwaren,60,L,,1.35,0,,1,,MwST-7
JA,680594, Müllbeutel ,mit Tragegriff,Haushaltwaren,35,L,,0.99,0,,1,,MwST-7
JA,868752, Müllbeutel,mit Tragegriff,Haushaltwaren,50,L,,0.55,0,,1,,MwST-7
```

**Pull request merged**